org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/ServerIssueUpdater.java

```diff
@@ -0,0 +1,120 @@
+import org.sonarsource.sonarlint.core.client.api.connected.ConnectedSonarLintEngine;
+import org.sonarsource.sonarlint.core.client.api.connected.ServerConfiguration;
+import org.sonarsource.sonarlint.core.client.api.connected.ServerIssue;
+import org.sonarsource.sonarlint.core.client.api.exceptions.DownloadException;
+
+public class ServerIssueUpdater {
+
+  private static final Logger LOGGER = new Logger();
+
+  private static final int THREADS_NUM = 5;
+  private static final int QUEUE_LIMIT = 20;
+
+  private final ExecutorService executorService;
+
+  private final IssueTrackerRegistry issueTrackerRegistry;
+
+  private final Console console = new Console();
+
+  public ServerIssueUpdater(IssueTrackerRegistry issueTrackerRegistry) {
+    final BlockingQueue<Runnable> queue = new ArrayBlockingQueue<>(QUEUE_LIMIT);
+    this.executorService = new ThreadPoolExecutor(THREADS_NUM, THREADS_NUM, 0L, TimeUnit.MILLISECONDS, queue);
+
+    this.issueTrackerRegistry = issueTrackerRegistry;
+  }
+
+  public void updateFor(ServerConfiguration serverConfiguration, ConnectedSonarLintEngine engine, String localModuleKey, String serverModuleKey, String relativePath) {
+    Runnable task = new IssueUpdateRunnable(serverConfiguration, engine, localModuleKey, serverModuleKey, relativePath);
+    try {
+      this.executorService.submit(task);
+    } catch (RejectedExecutionException e) {
+      LOGGER.debug("fetch and match server issues rejected for moduleKey=" + serverModuleKey + ", filepath=" + relativePath, e);
+    }
+  }
+
+  public void shutdown() {
+    List<Runnable> rejected = executorService.shutdownNow();
+    if (!rejected.isEmpty()) {
+      LOGGER.debug("rejected " + rejected.size() + " pending tasks");
+    }
+  }
+
+  private class IssueUpdateRunnable implements Runnable {
+    private final ServerConfiguration serverConfiguration;
+    private final ConnectedSonarLintEngine engine;
+    private final String localModuleKey;
+    private final String serverModuleKey;
+    private final String relativePath;
+
+    private IssueUpdateRunnable(ServerConfiguration serverConfiguration, ConnectedSonarLintEngine engine, String localModuleKey, String serverModuleKey, String relativePath) {
+      this.serverConfiguration = serverConfiguration;
+      this.engine = engine;
+      this.localModuleKey = localModuleKey;
+      this.serverModuleKey = serverModuleKey;
+      this.relativePath = relativePath;
+    }
+
+    @Override
+    public void run() {
+      try {
+        ConnectedSonarLintEngine.class.getProtectionDomain().getCodeSource().getLocation();
+        Iterator<ServerIssue> serverIssues = fetchServerIssues(serverConfiguration, engine, serverModuleKey, relativePath);
+        Collection<Trackable> serverIssuesTrackable = toStream(serverIssues).map(ServerIssueTrackable::new).collect(Collectors.toList());
+
+        issueTrackerRegistry.get(localModuleKey).matchAndTrackAsBase(relativePath, serverIssuesTrackable);
+      } catch (Throwable t) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Catch Exception instead of Throwable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1181)